### PR TITLE
Compiler support for scala-async, with a state machine transform phase running after erasure [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
[`scala-async`](https://github.com/scala/scala-async) ([SIP-22](https://docs.scala-lang.org/sips/async.html)) is currently implemented as a macro.

This pull request incorporates the bulk of the async transform into a fully blown compiler phase of the standard compiler. `scala-async` would remain as a simple macro that expands into the skeleton state machine class including adaptors for `scala.concurrent.Future`.

Other front end macros could also target the async phase adapting to other awaitable data types. The test cases include an [integration with  `java.util.concurrent.CompletableFuture`](https://github.com/scala/scala/pull/8816/files#diff-6fad4e47b4021c5655ebaa6fdfcf45c1R15). It should be also be straightforward to integrate with [`monix.Task`](https://monix.io/docs/2x/eval/task.html) and the like now, unlike [before](https://github.com/scala/scala-async/issues/149).

Why abandon the macro approach?

  - Typer is a somewhat unnatural place in the compiler pipeline for the expansion. The ANF transform is markedly simpler once things after `uncurry` and easier again after `erasure`. Custom integrations of async that use async pattern extactors and guards are _only_  possible when running after `patmat`.
  - Async translated blocks generate larger ASTs than the original source code. Such a a translation is better deferred to later in the pipeline for performance reasons.
  - Async leans heavily on compiler APIs, and some of these are unavailable int the macro API facade.

## Integrate Async into the compiler

  - Import the scala-async code base into `scala.tools.nsc.transform.async`
  - The front end macro is required to expand to `async(expr)(execCtx)` into (approximately) `{val temp = execCtx; class stateMachine { def apply(tr: Try) = { expr; () } }; new stateMachine.start()}`. Wrapping `expr` in a class lets intervening compiler phases do their job (e.g. captured enclosing values give rise to outer pointer usage).
  - The `async` phase proper runs after erasure. As before, it consists of a) a selective ANF transform of the expression; b) a translation of control flow that crosses async boundaries into separate states of a state machine; c) lifting locals that are referred to from multiple states into members of the state machine.
  - Refactor to be idiomatic compiler phase (use normal Transformers/Traversers APIs, use other internal APIs rather than `c.internal_`.
 - This functionality is opt-in with `-Xasync`. A new version of `scala-async` ([prototype](https://github.com/scala/scala-async/pull/237/files)) will be needed to take advantage.
 - Test cases have been imported into `test/async` and `junit/test/scala/tools/nsc/transform/async`.

## Improve the implementation of the ANF transform

  - Refactor the code for better performance and maintainability
  - Be more selective about when to transform code.
  - Simplify with the assumption that only post-erasure Tree shapes will be seen.

## Improve the implementation of the State Machine Transform

  - Refactor the code for better performance and maintainability
  - [Drastically reduce](https://gist.github.com/retronym/50839a051db9b6d1fcd0719d99ecbfc2) the number of states needed to represent pattern matches and if/else constructs with "sparse" usages of await among the branches. Blocks prior to async boundaries are now kept int the same state as the `If` or `LabelDef` itself.
  - Remove the restriction about using `await` in the operand of `Boolean#{||, &&}` and instead rewrite these to `If` trees.
 
## Improve performance

  - Compile time performance is [improved markedly](https://github.com/retronym/compiler-benchmark/pull/1). This is due to emitting fewer states, performing less bookkeeping, and avoiding or fixing inefficiencies in the way we interact with compiler facilities like tree attachments, tree transformers.
  - Runtime performance is likely improved for certain use cases, although the savings may be negligible compared to the cost of async boundaries for many realistic workloads. For each state transition that we avoid, we save one tableswitch, some null field assignments and read/writes of the state variable. Another benefit would be needing to lift fewer values into fields (saving space in the state machine and the put/get field cost.

## TODO
 - [x] Review/Refine the names of the methods in `AsyncStateMachine`
 - [ ] Longer term plan for how to avoid the need for the heuristic approach to deferring `@compileTimeOnly` await methods
 - [x] real world testing against code bases already usnig `scala-async`
 - [x] test coverage for:
   - [x] error cases in `UnsupportedAwaitAnalyzer`
   - [x] `if (tree.tpe =:= definitions.NothingTpe) {` branch in `transformMatchOrIf`
   - [x] `case td@TypeDef(_, _, tparams, rhs)               =>` in `Lifter` (maybe dead code)
   - [x] ThicketTransformer (add a unit test)
   - [x] `case Apply(fun, arg1 :: arg2 :: Nil) if isBooleanShortCircuit(fun.symbol) =>` in `AsyncTraverser`
  - [x] Fix positions of synthetic trees, the position of the apply method (the start of the async block or method) isn't suitable! Either use `NoPosition` of the preciding position. 
  - [ ] Test and update IntelliJ's debugger support for the changes to mangled names: https://github.com/retronym/intellij-scala/pull/1

### CODE REVIEW

**Comment by sjrd** [comment]
Can these changes create new shapes of trees using `LabelDef`s, notably in the interaction with pattern matching? That would potentially mean more work on Scala.js' side.

**Comment by retronym** [comment]
@sjrd Yes, I'm using label jumps like `continue` 

```scala
def apply(tr: Try[AnyRef]) = {
:while1
while (true) {
  state match {
     case <i> => 
        ...
          GOTO while1() // continue a new iteration of the dispatch loop
        ...
          GOTO case12() // forward jump to a `caseN` or `matchEnd` label that may
        ...
     case <j> =>
         :case12
     case _ => throw ...
  }
  GOTO while1()
}
```

The `GOTO while1` would need to be translated to a Javascript `continue`.

The `GOTO matchEnd` or `GOTO caseNext` labels now can target a label that is not in the same `Block`, which I expect would be problematic for your translation. These could be replaced with an extra pass (either by the async phase or by Scala.js) with `state = <state containing label>; while1()`.

I was also experimenting with a novel use of a `LabelDef` to encapsulate a few instructions used before each async boundary. Results: https://gist.github.com/retronym/c219f8759bd92b8f65804fbaeffcb6ec. That change isn't included in this PR though.

**Comment by sjrd** [comment]
We don't compile to JavaScript, but to our IR. Our IR doesn't have `continue`, but it has something much more fun: labeled blocks (see section 3.2.6 of [my thesis](https://lampwww.epfl.ch/~doeraene/thesis/)). From what you're showing, it seems we'll definitely need some more special casing to recognize that shape, but that we can translate it to the following IR:
```scala
while (true) {
  while1[void]: {
    val (newState, awaitable) = await[(int, Awaitable)]: {
      state match {
        case <i> =>
          ...
          return@while1 void
          ...
          return@await (j, future)
        case _ => ...
      }
      return@while1 void
    } // end await
    state = newState
    if (isComplete(awaitable))
      tr = awaitable.get
    else
      {awaitable.onComplete(this); return}
  } // end while1
}
```

**Comment by sjrd** [comment]
It seems I replied to an earlier version of your comment.

For `case` labels, we already have a solution, as long as their shape doesn't get more complicated because of the transform.

**Comment by sjrd** [comment]
I'd like to discuss how we can make this testable in Scala.js. At the moment it seems there are essentially two kinds of tests:
* partest tests that actually delegate to a kind of JUnit runner, with tests that systematically perform `Await.result`
* "driven" tests relying on a JUnit test that instantiates a compiler, then runs the compiled code via reflection

Neither kind will be testable in Scala.js, which becomes problematic now that the support of `async` can generate new shapes of trees. It means that we'll never be able to reliably test any of this in the context of Scala.js.

Could we instead use one way to write async tests that is (or can be made) compatible with Scala.js? Off the top of my head, I imagine we could have a single `AsyncTest` helper in `partest-extras`, that would like:
```scala
abstract class AsyncTest {
  def test: Future[Any]

  def main(args: Array[String]): Unit {
    Await.result(test, Duration.Inf)
  }
}
```
Then, *all* tests that need running would be written as partests that inherit from this class. In Scala.js, we would override that single class so that it doesn't use `Await.result`, but instead delegates to the event loop. That would be manageable, and would allow all these tests to be testable also on Scala.js.

**Comment by retronym** [comment]
@sjrd That sounds like a reasonable plan. 

The JUnit tests are written in that way because either:

  - they need to test the compiler plugin front-end that use `@autoasync`. However I could move this compiler plugin implementation into `partest-extras` and activate the plugin in selected partests with `-Xplugin`.
  - I added the tests there because it was more convenient for me to debug that an equivalent `partest` case. I can just move those back.

The partest cases can certainly be refactored to abstract over the `Await.result` call.

**Comment by retronym** [comment]
@sjrd I've converted one of the partest Junit tests to a reflection free (other than a `ClassTag) harness that doesn't directly use `Await.result`. Here's how it looks: https://github.com/scala/scala/compare/a29adccbbcea481abce38249dc54eda0b6baae72...retronym:topic/async-scalajs-test-compat?expand=1

Does this look okay?

**Comment by sjrd** [comment]
Thank you @retronym. It's still problematic because there is no way to implement
```scala
def block[T](f: Future[T]): T
```
on the JS platform.

Essentially each test must return a `Future[Unit]` instead of a `Unit`, and it must be the responsibility of `main` to to block on those futures (on the JVM) or to chain them through the even loop (in JS).

The method `test` would have the following signature:
```scala
  private val tests = mutable.ListBuffer[(String, () => Future[Unit])]()
  def test(name: String)(f: => Future[Unit]) = {
    tests += (name, () => f)
  }
```
and for example, the test that is currently:
```scala
    test("inlining block does not produce duplicate definition") {
      async {
        val f = 12
        val x = await(fut(f))

        {
          type X = Int
          val x: X = 42
          identity(x)
        }
        type X = Int
        x: X
      }.block
    }
```
should would lose the `.block` at the end.

A test that performs assertions after the `.block`, like
```scala
    test("inlining block in tail position does not produce duplicate definition") {
      async {
        val f = 12
        val x = await(fut(f))

        {
          val x = 42
          x
        }
      }.block mustBe (42)
```
would instead end in
```scala
      }.map { _ mustBe (42) }
```

**Comment by retronym** [comment]
@sjrd I've taken a middle ground by moving a subset of the test cases to use a synchronous DSL over `Option` rather than the asynchronous DSL over `s.c.Future`. These tests also don't use JUnit's reflection. The other tests are now segregated in `test/async/jvm`.

File: src/partest-extras/scala/tools/partest/async/Async.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  @compileTimeOnly("[async] `await` must be enclosed in an `async` block")
+  def await[T](completableFuture: Future[T]): T = ???
+
+  def impl(c: blackbox.Context)(body: c.Tree)(executionContext: c.Tree): c.Tree = {
+    import c.universe._
+    val awaitSym = typeOf[Async.type].decl(TermName("await"))
+    def mark(t: DefDef): Tree = {
+      c.internal.markForAsyncTransform(c.internal.enclosingOwner, t, awaitSym, Map.empty)
+    }
+    val name = TypeName("stateMachine$async_" + body.pos.line)
```
Comment (sjrd) [inline src/partest-extras/scala/tools/partest/async/Async.scala]: Could we avoid the `_` sign in the class name, please? That triggers a bail-out in the Scala.js back-end, which requires the use of Unicode characters in the generated JS identifiers. It works, of course, but it's inefficient (while generating the code) and ugly (in the generated .js file).

File: src/partest-extras/scala/tools/partest/async/AsyncStateMachine.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package scala.tools.partest.async
+
+// The async phase expects the state machine class to structurally conform to this interface.
+trait AsyncStateMachine[F, R] {
+  /** Assign `i` to the state variable */
+  protected def state$async_=(i: Int): Unit
+  /** Retrieve the current value of the state variable */
+  protected def state$async: Int
```
Comment (sjrd) [inline src/partest-extras/scala/tools/partest/async/AsyncStateMachine.scala]: Could we not use `$` in method names that users will have to see at some point? `$` are supposed to be reserved for the compiler, so we shouldn't encourage people by showing them us using such `$`s.

**Comment by sjrd** [comment]
@retronym Thank you. I managed to update the Scala.js backend enough to support the `tests/async/smoketest.scala`. See https://github.com/scala-js/scala-js/pull/4067 for the changes.

You said earlier:

> The `GOTO matchEnd` or `GOTO caseNext` labels now can target a label that is not in the same `Block`, which I expect would be problematic for your translation. These could be replaced with an extra pass (either by the async phase or by Scala.js) with `state = <state containing label>; while1()`.

I did not see any occurrence of this pattern in the smoke test. Consequently my updated Scala.js backend still does not support that. Is it really true that it can happen? If yes, could you give an example source code, ideally with `optionally`, that triggers this?

**Comment by retronym** [comment]
> I did not see any occurrence of this pattern in the smoke test. 

`matchEnd49` is defined and called from different `CaseDefs` here:

```
optionally {
      val x: Option[Either[Object, (String, String)]] = Some(Right(("a", "b")))
      x match {
        case Some(Left(_)) => 1
        case Some(Right(("a", "c"))) => 2
        case Some(Right(("a", "e"))) => 3
        case Some(Right(("a", x))) if "ab".isEmpty => 4
        case Some(Right(("a", "b"))) => value(Option(5))
        case Some(Right((y, x))) if x == y => 6
        case Some(Right((_, _))) => value(Option(7))
        case None => 8
      }
    }
```

```
        override def apply(tr$async: Option): Unit = while$(){
          try {
            stateMachine$async.this.state() match {
              case 0 => {
                val x: Option = new Some(scala.`package`.Right().apply(new Tuple2("a", "b")));
                <synthetic> var rc48: Boolean = false;
                <synthetic> <stable> var x2: Some = (null: Some);
                stateMachine$async.this.match$1 = 0;
                case <synthetic> val x1: Option = x;
                case50(){
                  if (x1.$isInstanceOf[Some]())
                    {
                      rc48 = true;
                      x2 = (x1.$asInstanceOf[Some](): Some);
                      {
                        <synthetic> val p3: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                        if (p3.$isInstanceOf[scala.util.Left]())
                          {
                            stateMachine$async.this.match$1 = 1;
                            matchEnd49()
                          }
                        else
                          case51()
                      }
                    }
                  else
                    case51();
                  ()
                };
                case51(){
                  if (rc48)
                    {
                      <synthetic> val p7: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                      if (p7.$isInstanceOf[scala.util.Right]())
                        {
                          <synthetic> val x8: scala.util.Right = (p7.$asInstanceOf[scala.util.Right](): scala.util.Right);
                          {
                            <synthetic> val p9: Tuple2 = x8.value().$asInstanceOf[Tuple2]();
                            if (p9.ne(null))
                              {
                                <synthetic> val p10: String = p9._1().$asInstanceOf[String]();
                                <synthetic> val p11: String = p9._2().$asInstanceOf[String]();
                                if ("a".==(p10).&&("c".==(p11)))
                                  {
                                    stateMachine$async.this.match$1 = 2;
                                    matchEnd49()
                                  }
                                else
                                  case52()
                              }
                            else
                              case52()
                          }
                        }
                      else
                        case52()
                    }
                  else
                    case52();
                  ()
                };
                case52(){
                  if (rc48)
                    {
                      <synthetic> val p13: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                      if (p13.$isInstanceOf[scala.util.Right]())
                        {
                          <synthetic> val x14: scala.util.Right = (p13.$asInstanceOf[scala.util.Right](): scala.util.Right);
                          {
                            <synthetic> val p15: Tuple2 = x14.value().$asInstanceOf[Tuple2]();
                            if (p15.ne(null))
                              {
                                <synthetic> val p16: String = p15._1().$asInstanceOf[String]();
                                <synthetic> val p17: String = p15._2().$asInstanceOf[String]();
                                if ("a".==(p16).&&("e".==(p17)))
                                  {
                                    stateMachine$async.this.match$1 = 3;
                                    matchEnd49()
                                  }
                                else
                                  case53()
                              }
                            else
                              case53()
                          }
                        }
                      else
                        case53()
                    }
                  else
                    case53();
                  ()
                };
                case53(){
                  if (rc48)
                    {
                      <synthetic> val p19: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                      if (p19.$isInstanceOf[scala.util.Right]())
                        {
                          <synthetic> val x20: scala.util.Right = (p19.$asInstanceOf[scala.util.Right](): scala.util.Right);
                          {
                            <synthetic> val p21: Tuple2 = x20.value().$asInstanceOf[Tuple2]();
                            if (p21.ne(null))
                              {
                                <synthetic> val p22: String = p21._1().$asInstanceOf[String]();
                                if ("a".==(p22).&&("ab".isEmpty()))
                                  {
                                    stateMachine$async.this.match$1 = 4;
                                    matchEnd49()
                                  }
                                else
                                  case54()
                              }
                            else
                              case54()
                          }
                        }
                      else
                        case54()
                    }
                  else
                    case54();
                  ()
                };
                case54(){
                  if (rc48)
                    {
                      <synthetic> val p24: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                      if (p24.$isInstanceOf[scala.util.Right]())
                        {
                          <synthetic> val x25: scala.util.Right = (p24.$asInstanceOf[scala.util.Right](): scala.util.Right);
                          <synthetic> val p26: Tuple2 = x25.value().$asInstanceOf[Tuple2]();
                          if (p26.ne(null))
                            {
                              <synthetic> val p27: String = p26._1().$asInstanceOf[String]();
                              <synthetic> val p28: String = p26._2().$asInstanceOf[String]();
                              if ("a".==(p27).&&("b".==(p28)))
                                {
                                  val awaitable$async: Option = scala.Option.apply(scala.Int.box(5));
                                  tr$async = stateMachine$async.this.getCompleted(awaitable$async);
                                  stateMachine$async.this.state_=(1);
                                  if (null.!=(tr$async))
                                    while$()
                                  else
                                    {
                                      stateMachine$async.this.onComplete(awaitable$async);
                                      return ()
                                    }
                                }
                              else
                                {
                                  case55();
                                  ()
                                }
                            }
                          else
                            {
                              case55();
                              ()
                            }
                        }
                      else
                        {
                          case55();
                          ()
                        }
                    }
                  else
                    {
                      case55();
                      ()
                    }
                };
                case55(){
                  if (rc48)
                    {
                      <synthetic> val p30: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                      if (p30.$isInstanceOf[scala.util.Right]())
                        {
                          <synthetic> val x31: scala.util.Right = (p30.$asInstanceOf[scala.util.Right](): scala.util.Right);
                          {
                            <synthetic> val p32: Tuple2 = x31.value().$asInstanceOf[Tuple2]();
                            if (p32.ne(null))
                              {
                                val y: String = p32._1().$asInstanceOf[String]();
                                val x: String = p32._2().$asInstanceOf[String]();
                                if (x.==(y))
                                  {
                                    stateMachine$async.this.match$1 = 6;
                                    stateMachine$async.this.state_=(3);
                                    while$()
                                  }
                                else
                                  case56()
                              }
                            else
                              case56()
                          }
                        }
                      else
                        case56()
                    }
                  else
                    case56();
                  ()
                };
                case56(){
                  if (rc48)
                    {
                      <synthetic> val p34: scala.util.Either = x2.value().$asInstanceOf[scala.util.Either]();
                      if (p34.$isInstanceOf[scala.util.Right]())
                        {
                          <synthetic> val x35: scala.util.Right = (p34.$asInstanceOf[scala.util.Right](): scala.util.Right);
                          <synthetic> val p36: Tuple2 = x35.value().$asInstanceOf[Tuple2]();
                          if (p36.ne(null))
                            {
                              val awaitable$async: Option = scala.Option.apply(scala.Int.box(7));
                              tr$async = stateMachine$async.this.getCompleted(awaitable$async);
                              stateMachine$async.this.state_=(2);
                              if (null.!=(tr$async))
                                while$()
                              else
                                {
                                  stateMachine$async.this.onComplete(awaitable$async);
                                  return ()
                                }
                            }
                          else
                            {
                              case57();
                              ()
                            }
                        }
                      else
                        {
                          case57();
                          ()
                        }
                    }
                  else
                    {
                      case57();
                      ()
                    }
                };
                case57(){
                  if (scala.None.==(x1))
                    {
                      stateMachine$async.this.match$1 = 8;
                      stateMachine$async.this.state_=(3);
                      while$()
                    }
                  else
                    case58();
                  ()
                };
                case58(){
                  throw new MatchError(x1)
                };
                stateMachine$async.this.state_=(3)
              }
              case 1 => {
                <synthetic> val await$1: Object = {
                  val tryGetResult$async: Object = stateMachine$async.this.tryGet(tr$async);
                  if (stateMachine$async.this.==(tryGetResult$async))
                    return ()
                  else
                    tryGetResult$async.$asInstanceOf[Object]()
                };
                <synthetic> val x$1: Int = unbox(await$1);
                stateMachine$async.this.match$1 = x$1;
                stateMachine$async.this.state_=(3);
                while$()
              }
              case 2 => {
                <synthetic> val await$2: Object = {
                  val tryGetResult$async: Object = stateMachine$async.this.tryGet(tr$async);
                  if (stateMachine$async.this.==(tryGetResult$async))
                    return ()
                  else
                    tryGetResult$async.$asInstanceOf[Object]()
                };
                <synthetic> val x$2: Int = unbox(await$2);
                stateMachine$async.this.match$1 = x$2;
                stateMachine$async.this.state_=(3);
                while$()
              }
              case 3 => {
                matchEnd49(){
                  ()
                };
                stateMachine$async.this.completeSuccess(scala.Int.box(stateMachine$async.this.match$1));
                return ()
              }
              case 4 => {
                ()
              }
              case _ => throw new IllegalStateException(java.lang.String.valueOf(stateMachine$async.this.state()))
            }
          } catch {
            case (throwable$async @ (_: Throwable)) => {
              stateMachine$async.this.completeFailure(throwable$async);
              return ()
            }
          };
          while$()
        }
      };
```

**Comment by sjrd** [comment]
I see. I could indeed reproduce that behavior.

The bad news is: I have no idea how to deal with this in the Scala.js back-end. If you look at the PR I mentioned above, https://github.com/scala-js/scala-js/pull/4067, and in particular at the commit "Generalize the codegen of `LabelDef`s for pattern matching", you can see the places where with deal with patmat-generated `LabelDef`s. You'll see that with the changes of that PR, we can in general deal with the following shapes:
```scala
  {
    prologue
    case1() {
      ...
        matchEnd(1)
      ...
        case2()
    }
    case2() {
      ...
    }
    ...
    caseN() {
      ...
    }
    <optional>:
      matchEnd(x) {
        ...
      }
    rest
  }
```
which we translate as:
```scala
// Scala.js IR
val x = matchEnd[int]: {
  case2[void]: {
    ...
      return@matchEnd 1
    ...
      return@case2 undefined
  }
  case3[void]: {
    ...
  }
  caseN[void]: {
    ...
  }
  ... // content of caseN() before
}
x
```
We can do this because all jumps are at least to a label that is defined in an enclosing scope of where the jump happens.

Now with the new async transformation, it seems we can have a jump in one scope calling a LabelDef defined in an unrelated scope. It is very problematic for Scala.js, but honestly I also find that a bit terrifying in terms of sanity of scalac Trees.

You suggested that instead of having a jump to `matchEnd()`, we could have something like
```scala
async_70$1.this.match$2 = 2;
async_70$1.this.state$async_=(3);
while$()
```
Recovering that after the fact seems like a daunting task. How would we identify those places? How would we make sure that we only alter async-generated stuff, and not other code? It seems completely unwieldy.

**Couldn't the async transform directly generate the above?** That would avoid having jumps to labels that are not declared in the scope of the jumps, which seems like a decent property to have, even regardless of Scala.js.

The only concern I can think of is that it would generate less-than-optimal bytecode on the JVM. But that seems to already happen in other places, for example in the `case 2 =>` above.

It seems to me that this would be the best compromise between the requirements of the two platforms.

Another reason is that the schema wouldn't be portable to Dotty. Dotty has exactly the same notion of labeled blocks as the Scala.js IR, so we would have to generate something compatible with those `Labeled` blocks anyway.

**Comment by retronym** [comment]
I've added a post-processing pass to the `async` phase to find residual label jumps that cross state transitions and replace them with `state = ...; while()`.

**Comment by sjrd** [comment]
> I've added a post-processing pass to the async phase to find residual label jumps that cross state transitions and replace them with state = ...; while().

Thank you! This is perfect. The snippet from above worked out of the box with your latest changes. :)

Thanks also for addressing my other comments :)

File: src/reflect/scala/reflect/api/Internals.scala [hunk 3]
```diff
@@ -374,6 +374,8 @@ trait Internals { self: Universe =>
      */
     def boundedWildcardType(bounds: TypeBounds): BoundedWildcardType
 
+    def markForAsyncTransform(owner: Symbol, method: DefDef, awaitSymbol: Symbol, config: Map[String, AnyRef]): DefDef = method
```
Comment (lrytz) [inline src/reflect/scala/reflect/api/Internals.scala:388]: Would be good to have some comment here; this is the entry point for defining custom async-style macros, right? e.g. https://github.com/retronym/monad-ui/blob/master/src/main/scala/monadui/OutputAsyncDsl.scala

File: src/compiler/scala/tools/nsc/Global.scala [hunk 4]
```diff
@@ -568,11 +575,17 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     val runsRightAfter = Some("erasure")
   } with PostErasure
 
+  // phaseName = "async"
+  object async extends {
+    val global: Global.this.type = Global.this
+    val runsAfter = List("posterasure")
+    val runsRightAfter = Some("posterasure")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/Global.scala]: is the hard link here required, or coud it be `None`?

File: src/compiler/scala/tools/nsc/Global.scala [hunk 5]
```diff
@@ -1336,6 +1350,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     def runIsAt(ph: Phase)   = globalPhase.id == ph.id
     def runIsAtOptimiz       = runIsAt(jvmPhase)
 
+    firstPhase.iterator.foreach(_.init())
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/Global.scala]: Leftover? It seems there's no phase that does something in `init()`.

File: src/compiler/scala/tools/nsc/Global.scala [hunk 6]
```diff
@@ -999,7 +1013,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
        definitions.isDefinitionsInitialized
     && rootMirror.isMirrorInitialized
   )
-  override def isPastTyper = isPast(currentRun.typerPhase)
+  override def isPastTyper = globalPhase != null && isPast(currentRun.typerPhase)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/Global.scala]: I don't find any assignments of `null` to `globalPhase`?
```
  var globalPhase: Phase = NoPhase
```

**Comment by sjrd** [comment]
With a few more changes on Scala.js' side, I could make all the tests in edge-cases pass :)

I noticed that, with the latest state of this PR, (almost) all generated switch'es have a last state that is empty dead code, and that would loop forever if reached. For example, the very first `testBasic()` in `smoketest.scala` generates:
```scala
  final class Test$stateMachine$async$1 extends scala.tools.partest.async.OptionStateMachine {
    <synthetic> <stable> private[this] var await$1: Object = _;
    override def apply(tr$async: Option): Unit = while$(){
      try {
        Test$stateMachine$async$1.this.state() match {
          case 0 => {
            val awaitable$async: Some = new Some(scala.Int.box(1));
            tr$async = Test$stateMachine$async$1.this.getCompleted(awaitable$async);
            Test$stateMachine$async$1.this.state_=(1);
            if (null.!=(tr$async))
              while$()
            else
              {
                Test$stateMachine$async$1.this.onComplete(awaitable$async);
                return ()
              }
          }
          case 1 => {
            {
              val tryGetResult$async: Object = Test$stateMachine$async$1.this.tryGet(tr$async);
              if (Test$stateMachine$async$1.this.==(tryGetResult$async))
                Test$stateMachine$async$1.this.await$1 = return ()
              else
                Test$stateMachine$async$1.this.await$1 = tryGetResult$async.$asInstanceOf[Object]()
            };
            val awaitable$async: Some = new Some(scala.Int.box(2));
            tr$async = Test$stateMachine$async$1.this.getCompleted(awaitable$async);
            Test$stateMachine$async$1.this.state_=(2);
            if (null.!=(tr$async))
              while$()
            else
              {
                Test$stateMachine$async$1.this.onComplete(awaitable$async);
                return ()
              }
          }
          case 2 => {
            <synthetic> val await$2: Object = {
              val tryGetResult$async: Object = Test$stateMachine$async$1.this.tryGet(tr$async);
              if (Test$stateMachine$async$1.this.==(tryGetResult$async))
                return ()
              else
                tryGetResult$async.$asInstanceOf[Object]()
            };
            <synthetic> val x$1: Int = scala.Int.unbox(await$2);
            <synthetic> val x$2: Int = scala.Int.unbox(Test$stateMachine$async$1.this.await$1).+(x$1);
            Test$stateMachine$async$1.this.completeSuccess(scala.Int.box(x$2));
            return ();
            Test$stateMachine$async$1.this.await$1 = null
          }
          case 3 => {
            ()
          }
          case _ => throw new IllegalStateException(java.lang.String.valueOf(Test$stateMachine$async$1.this.state()))
        }
      } catch {
        case (throwable$async @ (_: Throwable)) => {
          Test$stateMachine$async$1.this.completeFailure(throwable$async);
          return ()
        }
      };
      while$()
    };
    def <init>(): Test$stateMachine$async$1 = {
      Test$stateMachine$async$1.super.<init>();
      ()
    }
  }
```
Note the `case 3 => { () }`. If reached, it would loop forever since it doesn't change the state, but no other code sets the state to 3. The same happens in all async blocks I looked at.

This is not wrong/harmful; it's just useless trees.

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 7]
```diff
@@ -327,10 +327,15 @@ trait Implicits {
    */
   def memberWildcardType(name: Name, tp: Type) = {
     val result = refinedType(List(WildcardType), NoSymbol)
-    name match {
-      case x: TermName => result.typeSymbol.newMethod(x) setInfoAndEnter tp
-      case x: TypeName => result.typeSymbol.newAbstractType(x) setInfoAndEnter tp
+    val owner = result.typeSymbol orElse { // TODO async (when after erasure?)...
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: any insight why implicit search triggers for code generated by async?

File: src/compiler/scala/tools/reflect/FastTrack.scala [hunk 8]
```diff
@@ -44,6 +44,9 @@ class FastTrack[MacrosAndAnalyzer <: Macros with Analyzer](val macros: MacrosAnd
   private def makeWhitebox(sym: Symbol)(pf: PartialFunction[Applied, MacroContext => Tree]) =
     sym -> new FastTrackEntry(pf, isBlackbox = false)
 
+  private def makeBlackBoxIfExists(sym_pf: (Symbol, PartialFunction[Applied, MacroContext => Tree])) =
```
Comment (lrytz) [inline src/compiler/scala/tools/reflect/FastTrack.scala]: unused?

File: src/reflect/scala/reflect/internal/StdAttachments.scala [hunk 9]
```diff
@@ -119,4 +119,9 @@ trait StdAttachments {
   case object KnownDirectSubclassesCalled extends PlainAttachment
 
   class QualTypeSymAttachment(val sym: Symbol)
+
+  // When typing a Def with this attachment, change the owner of its RHS from origalOwner to the symbol of the Def
+  case class ChangeOwnerAttachment(originalOwner: Symbol)
+
+  case object SuppressPureExpressionWarning extends PlainAttachment
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/StdAttachments.scala]: Unused?

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 10]
```diff
@@ -478,7 +478,14 @@ trait Trees extends api.Trees {
   case class ArrayValue(elemtpt: Tree, elems: List[Tree]) extends TermTree
 
   case class Function(vparams: List[ValDef], body: Tree)
-       extends SymTree with TermTree with FunctionApi
+       extends SymTree with TermTree with FunctionApi {
+
+    override def tpe_=(tp: Type): Unit = {
+      if (!definitions.isFunctionType(tp))
+        new Throwable().printStackTrace()
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/Trees.scala]: remove debug code?

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 11]
```diff
@@ -1531,6 +1538,7 @@ trait Trees extends api.Trees {
   }
 
   class ChangeOwnerTraverser(val oldowner: Symbol, val newowner: Symbol) extends Traverser {
+    assert(newowner != NoSymbol, oldowner)
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/Trees.scala]: remove, or extend message?

**Review by lrytz — COMMENTED** [review]
A few minor comments. I also pushed a commit to remove (hopefully) unused code.

File: src/compiler/scala/tools/nsc/Global.scala (hunk 6 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/Global.scala]: Removed, this must have come up when I was trying the `Phase.init()` hook.

File: src/reflect/scala/reflect/api/Internals.scala (hunk 3 above)
Comment (retronym) [reply src/reflect/scala/reflect/api/Internals.scala:388]: Comments added.

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala (hunk 7 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: Doesn't seem to happen anymore, probably a leftover from the initial state of the PR in 6a665e97f991c1fe039fdb269c5b26399967ccaa or an even earlier version. Removed

File: src/compiler/scala/tools/nsc/Global.scala (hunk 4 above)
Comment (retronym) [reply src/compiler/scala/tools/nsc/Global.scala]: `None` seems fine as `lambdalift` asks that it `runsAfter` `async`.

**Review by lrytz — APPROVED** [review]

**Comment by lrytz** [comment]
About the risk for regressions that this change may cause:
  - All non-trivial changes in this PR are behind a new `-Xasync` compiler flag (for now)
  - Existing scala-async releases will continue to work exactly the same way with compilers that include this change

**Pull request merged**